### PR TITLE
Write merkle root of deploys' approvals to global state after block execution

### LINKED ISSUE #3234: Put deploy approvals Merkle tree root into the global state after each block execution.
To allow proving which deploy approvals were used for execution, add the root of the Merkle tree whose leaves are the deploy approval hashes into the global state after each block execution. The key should be based on the current block height

### PR DESCRIPTION
**Pull request by Fraser999**
This PR introduces writing a Merkle root of deploy approvals after block execution, in order to allow later validation of these approvals.

Closes #3234.

### CODE REVIEW

**Review by afck — APPROVED** [review]
Looks good to me! (But someone else should approve, since I wrote part of it.)

File: execution_engine_testing/tests/src/test/explorer/faucet.rs [hunk 1]
```diff
@@ -814,10 +814,10 @@ fn faucet_costs() {
...
     // without understanding why the cost has changed.  If the costs do change, it should be
     // reflected in the "Costs by Entry Point" section of the faucet crate's README.md.
-    const EXPECTED_FAUCET_INSTALL_COST: u64 = 73_977_063_780;
-    const EXPECTED_FAUCET_SET_VARIABLES_COST: u64 = 48_195_450;
-    const EXPECTED_FAUCET_CALL_BY_INSTALLER_COST: u64 = 2_584_138_930;
-    const EXPECTED_FAUCET_CALL_BY_USER_COST: u64 = 2_530_886_950;
+    const EXPECTED_FAUCET_INSTALL_COST: u64 = 74_041_335_220;
+    const EXPECTED_FAUCET_SET_VARIABLES_COST: u64 = 48_201_170;
+    const EXPECTED_FAUCET_CALL_BY_INSTALLER_COST: u64 = 2_584_143_770;
+    const EXPECTED_FAUCET_CALL_BY_USER_COST: u64 = 2_530_897_950;
```
Comment (goral09) [inline execution_engine_testing/tests/src/test/explorer/faucet.rs]: Why did the costs change? Is it b/c we're writing the additional value?

File: node/CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ All notable changes to this project will be documented in this file.  The format
 
 ### Added
 * Introduce fast-syncing to join the network, avoiding the need to execute every block to catch up.
+* Merkle root hashes for execution results and deploy approvals are written to global state after each block execution.
```
Comment (goral09) [inline node/CHANGELOG.md:18]: I don't think your PR also writes the merkle root of execution results.

Comment (goral09) [reply node/CHANGELOG.md:18]: Ah, it was added in a previous PR but changelog wasn't updated :+1:

**Review by goral09 — APPROVED** [review]

**Comment by Fraser999** [comment]
bors merge

File: execution_engine_testing/tests/src/test/explorer/faucet.rs (hunk 1 above)
Comment (Fraser999) [reply execution_engine_testing/tests/src/test/explorer/faucet.rs]: I believe it's down to the extra variant on the `Key` type.

**Comment by Fraser999** [comment]
bors merge

**Review by fizyk20 — APPROVED** [review]
LGTM :+1:

File: node/src/components/contract_runtime/operations.rs [hunk 3]
```diff
@@ -100,36 +100,40 @@ pub fn execute_finalized_block(
...
     }
 
-    let execution_results_root_hash = compute_execution_results_root_hash(
-        &mut execution_results.iter().map(|(_, _, result)| result),
-    )?;
+    // Write the deploy approvals and execution results Merkle root hashes to global state if there
+    // were any deploys.
     let block_height = finalized_block.height();
+    if let Some(deploy_approvals_root_hash) = maybe_deploy_approvals_root_hash {
+        let execution_results_root_hash = compute_execution_results_root_hash(
```
Comment (goral09) [inline node/src/components/contract_runtime/operations.rs:107]: Why not check for the the block effects as well?

Comment (Fraser999) [reply node/src/components/contract_runtime/operations.rs:107]: We only call `compute_execution_results_root_hash` if the set of deploys is non-empty.  I guess I could print an error in there if we pass an empty set of deploys, but seems low risk to me.

**Review by goral09 — APPROVED** [review]

**Review by fizyk20 — APPROVED** [review]

**Review by goral09 — APPROVED** [review]

**Comment by Fraser999** [comment]
bors merge

**Pull request merged**